crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -850,6 +927,40 @@ impl PgManager {
             .transpose()
     }
 
+    async fn multi_get_events(
+        &self,
+        first: Option<u64>,
+        after: Option<String>,
+        last: Option<u64>,
+        before: Option<String>,
+        filter: Option<EventFilter>,
+    ) -> Result<Option<(Vec<StoredEvent>, bool)>, Error> {
+        let descending_order = last.is_some();
```

Comment:
Not the first time I've failed to meet specs. I'll remove this, and fix the other connections in a separate PR after reviewing the specs

Reply:
Fixing this is turning out to be more involved than expected. Any 'after' cursor requires reversing the order of a subselect. For example, last 5 after E ASC is the sub query first 10 before E DESC, order by ASC to get VWXYZ 

This is not challenging to implement in raw SQL, it's quite simple really, 
but it is complicated by the fact that diesel does not have an OOTB representation for subselects in the dsl query builder.

The approaches and issues I ran into are outlined below. We could also consider doing the ordering on the application, although it sucks that getting subselects to work is so difficult. But if that's alright then I can avoid the headache below. 

In any case, do you think we could approve the changes here (I will add the querying logic for generic event types) so people can use combination filters, and I'll continue to work on fixing the connection logic separately? 

More details: I've tried multiple approaches, like
1. Following my implementation of a custom Explain query and creating a Subquery to represent the SQL logic. It wraps the original query in a `SELECT * FROM (original query) AS SUB`. The issue is that the modified QueryFragment is used only if we call LoadQuery on Subquery, and I haven't figured out how to 

the order by logic is unique for each entity, and I haven't figured out a way to allow additional filters when Subquery is applied. A workaround is to create some OrderedSubquery type for each entity, and then write the logic in raw sql ... 
2. Assuming that I can get the custom type to work, it still needs to play well with Explain, and I haven't figured out a way to turn Subquery or a custom query type into a BoxedQuery

My vision is we could do something like:
```
query.filter....subselect(all_fields).order(tx_sequence_number.asc())
```
3. Combining raw sql

Reply:
Because diesel does not have a subselect ootb we need to extend dsl querybuilder. I was hoping to do something similar to the implementation of the custom `Explain` query. Create a custom `Subquery` that implements `QueryFragment` to wrap the original query in a `SELECT * FROM (the original query) AS SUB`. However, I run into a few issues when doing so:

The `Explain` query is the final state of the query before execution. But after we produce the subquery, we want to apply the final order by condition. The implementation for `QueryFragment` only kicks in if we run `LoadQueryDsl` on that `Subquery`. The type of the query goes from `BoxedSelectStatement` to `Subqueried<BoxedSelectStatement>`. Ideally though, Diesel sees this as a BoxedSelectStatement still, so we can continue to work with it.

Alternatively, I can implement a `Subquery` for every table type we have, although that means we'll need to update all the existing table::BoxedQuery types to this `Subquery` and make `Explain` work with the custom type.